### ISSUE TITLE
Teensy battery emulator with driver

### ISSUE THREAD

**Issue by forderud** [forderud/BatterySimulator#23]
Alternative to #33 for battery emulation.

![image](https://github.com/user-attachments/assets/e4f7840c-e9f9-46dc-83ea-64ecf5e80ee3)

### Idea
* [Teensy](https://www.pjrc.com/teensy/) 4.x board that exposes a USB device with simulated battery pack status values that vary over time.
* Battery miniport driver to receive battery status events and forward them to Windows

### USB protocol
1. Header: Send `BATTERY\0` (8bytes), [`BATTERY_INFORMATION`](https://learn.microsoft.com/en-us/windows/win32/power/battery-information-str) struct (36bytes) and a [`BATTERY_STATUS`](https://learn.microsoft.com/en-us/windows/win32/power/battery-status-str) struct (16bytes) to host,
2. Afterwards, send `BATTERY_STATUS` structs (16bytes) on-demand to host when the status changes.

USB pipe: Use `IN interrupt` endpoint (max 64byte package size, max 255ms polling interval)

### Teensy firmware
Teensy dev setup: https://www.pjrc.com/teensy/td_download.html

Implement the protocol above and transmit battery charge changes periodically.

Open questions:
* USB VID & PID: https://devicehunt.com/view/type/usb/vendor/16C0

### Battery driver
`Class=Battery`  - can most likely be combined with USB client driver

Notes:
* [Overview of developing Windows client drivers for USB devices](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-driver-development-guide)
* Might need to use [continuous reader](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/how-to-use-the-continous-reader-for-getting-data-from-a-usb-endpoint--umdf-) for polling the USB device.
* Sample: [kmdf_fx2](https://github.com/microsoft/Windows-driver-samples/tree/main/usb/kmdf_fx2) - uses continuous reader to read from an interrupt endpoint.
* [`WdfUsbTargetPipeReadSynchronously`](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdfusb/nf-wdfusb-wdfusbtargetpipereadsynchronously) for blocking read from interrupt endpoint during initialization

## Implementation stages
1. Write Teensy firmware
1. Write a simple USB client driver that exposes `IOCTL_BATTERY_QUERY_INFORMATION` and  `IOCTL_BATTERY_QUERY_STATUS` to user-space.
1. Merge USB driver with `simbatt` sample

## Links
* https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-filter-driver-for-supporting-chargers
* https://www.usb.org/document-library/battery-charging-v12-spec-and-adopters-agreement

**Comment by forderud** [comment]
## hidbatt.inf content:
```

[Version]
Signature="$WINDOWS NT$"
Class=Battery
ClassGuid={72631e54-78a4-11d0-bcf7-00aa00b7b32a}
Provider=%MSFT%
DriverVer = 06/21/2006,10.0.19041.1

[SourceDisksNames]
3426=windows cd

[SourceDisksFiles]
hidbatt.sys = 3426

[DestinationDirs]
DefaultDestDir = 12

[ControlFlags]
BasicDriverOk = *
ExcludeFromSelect = *

[Manufacturer]
%MSFT%=Microsoft,NTamd64

[Microsoft.NTamd64]
%HID_DEVICE_BATTERY.DeviceDesc% = HidBatt_Inst,HID_DEVICE_UP:0084_U:0004

[HidBatt_Inst]
CopyFiles = HidBattInstCopyFiles
Include = battery.inf
Needs = Battery_Inst

[HidBattInstCopyFiles]
hidbatt.sys,,,0x100

[HidBatt_Inst.Services]
AddService=HidBatt,%SPSVCINST_ASSOCSERVICE%,HidBatt_Service_Inst

[HidBatt_Service_Inst]
DisplayName    = %HidBatt.SvcDesc%
ServiceType    = %SERVICE_KERNEL_DRIVER%
StartType      = %SERVICE_DEMAND_START%
ErrorControl   = %SERVICE_ERROR_NORMAL%
ServiceBinary  = %12%\HidBatt.sys

[strings]
;Localizable Strings
MSFT                     = "Microsoft"
HID_DEVICE_BATTERY.DeviceDesc = "HID UPS Battery"
HidBatt.SvcDesc          = "HID UPS Battery Driver"

;Non-Localizable Strings
SPSVCINST_ASSOCSERVICE   = 0x00000002
SERVICE_KERNEL_DRIVER    = 1
SERVICE_DEMAND_START     = 3
SERVICE_ERROR_NORMAL     = 1
```

**Comment by forderud** [comment]
Closed in favor of #33
